src/hexen/p_enemy.c

```diff
@@ -539,15 +553,19 @@ boolean P_LookForPlayers(mobj_t * actor, boolean allaround)
     }
     c = 0;
 
-    // NOTE: This behavior has been changed from the Vanilla behavior, where
-    // an infinite loop can occur if players 0-3 all quit the game. Although
-    // technically this is not what Vanilla does, fixing this is highly
-    // desirable, and having the game simply lock up is not acceptable.
-    // stop = (actor->lastlook - 1) & 3;
-    // for (;; actor->lastlook = (actor->lastlook + 1) & 3)
-
-    stop = (actor->lastlook + maxplayers - 1) % maxplayers;
-    for (;; actor->lastlook = (actor->lastlook + 1) % maxplayers)
+    if (demorecording || demoplayback)
+    {
+        // NOTE: This behaviour can cause an infinite loop in vanilla
+        // if players 0-3 all quit the game.
+        // However, this is essential for maintaining demo compatibility,
+        // so we must use the "broken" behaviour when demos are involved.
+        stop = (actor->lastlook - 1) & 3;
+    }
+    else
+    {
+        stop = (actor->lastlook + maxplayers - 1) % maxplayers;
```

Comment:
If the `if()` case fixes demo playback, why do we want to do this `else` case when demos are not involved? Is the fix here just wrong?

Reply:
Your original comment / fix (https://github.com/chocolate-doom/chocolate-doom/commit/d6693cc9e17fdbee7033e47b7b8da249da118d8f) says that the game can get stuck in an infinite loop here. The else condition is that "fixed" behaviour, in that it doesn't have a risk of infinite loops. But that change also violated demo compatibility, so it can't be used while demos are involved.

Reply:
If your new behavior fixes demo compatibility, I think that's the vanilla behavior. So it seems like my fix was wrong and we should just use your new behavior.

Although: if vanilla crashes/locks up here, is it more accurate to say that neither behaviors are truly vanilla? Are we just trying to maintain "compatibility" with demos recorded in other source ports? That may be fine given the circumstances, but we should probably protect the change using the `D_NonVanillaRecord` and `D_NonVanillaPlayback` functions.

Reply:
The demos in question were recorded in an original environment (hexen+ or otherwise), so this is compatibility with the original engine. I'd vote for just using the implementation with `3` and not the one using `maxplayers`, as that's the original code as far as I can tell.

Reply:
Couldn't we just `break` the loop if `(actor->lastlook + 1) & 3 == actor->lastlook` to retain Vanilla behaviour and avoid the infinite loop at the same time?